src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,934 @@
+                    fixed_columns.insert(maybe_fixed_column);
+
+                    const ActionsDAG::Node * maybe_injective = maybe_fixed_column;
+                    while (maybe_injective->type == ActionsDAG::ActionType::FUNCTION
+                        && maybe_injective->children.size() == 1
+                        && maybe_injective->function_base->isInjective({}))
+                    {
+                        maybe_injective = maybe_injective->children.front();
+                        fixed_columns.insert(maybe_injective);
+                    }
+                }
+            }
+        }
+    }
+}
+
+void appendExpression(ActionsDAGPtr & dag, const ActionsDAGPtr & expression)
+{
+    if (dag)
+        dag->mergeInplace(std::move(*expression->clone()));
+    else
+        dag = expression->clone();
+}
+
+void buildSortingDAG(QueryPlan::Node & node, ActionsDAGPtr & dag, FixedColumns & fixed_columns)
+{
+    IQueryPlanStep * step = node.step.get();
+    if (auto * reading = typeid_cast<ReadFromMergeTree *>(step))
+    {
+        if (const auto * prewhere_info = reading->getPrewhereInfo())
+        {
+            if (prewhere_info->prewhere_actions)
+            {
+                //std::cerr << "====== Adding prewhere " << std::endl;
+                appendExpression(dag, prewhere_info->prewhere_actions);
+                if (const auto * filter_expression = dag->tryFindInOutputs(prewhere_info->prewhere_column_name))
+                    appendFixedColumnsFromFilterExpression(*filter_expression, fixed_columns);
+            }
+        }
+        return;
+    }
+
+    if (node.children.size() != 1)
+        return;
+
+    buildSortingDAG(*node.children.front(), dag, fixed_columns);
+
+    if (auto * expression = typeid_cast<ExpressionStep *>(step))
+        appendExpression(dag, expression->getExpression());
+
+    if (auto * filter = typeid_cast<FilterStep *>(step))
+    {
+        appendExpression(dag, filter->getExpression());
+        if (const auto * filter_expression = dag->tryFindInOutputs(filter->getFilterColumnName()))
+            appendFixedColumnsFromFilterExpression(*filter_expression, fixed_columns);
+    }
+
+    if (auto * array_join = typeid_cast<ArrayJoinStep *>(step))
+    {
+        const auto & array_joined_columns = array_join->arrayJoin()->columns;
+
+        /// Remove array joined columns from outputs.
+        /// Types are changed after ARRAY JOIN, and we can't use this columns anyway.
+        ActionsDAG::NodeRawConstPtrs outputs;
+        outputs.reserve(dag->getOutputs().size());
+
+        for (const auto & output : dag->getOutputs())
+        {
+            if (!array_joined_columns.contains(output->result_name))
+                outputs.push_back(output);
+        }
+    }
+}
+
+/// Add more functions to fixed columns.
+/// Functions result is fixed if all arguments are fixed or constants.
+void enreachFixedColumns(const ActionsDAG & dag, FixedColumns & fixed_columns)
```

Comment:
Consider to name function `collectFixedColumns`.

Reply:
Do not agree here. Cause we only add a few more columns based on existing set.